### DISCUSSION TITLE
Problem with shims due to 10.9 target

### DISCUSSION THREAD

**Discussion by Gcenx** [Homebrew/discussions#2980] [Everyday usage] [answered]
I’m not sure if this would found for an actual issue for brew but let me explain

Xcode10 made a change blocking the useable of `-no_new_main` unless `-mmacosx-version-min=10.7` (or
earlier) is used. brew shims are setting 
`-mmacosx-version-min=10.9` https://github.com/Homebrew/brew/commit/481ffcf4e105916a2a24c44d5f28c8dff39bcfd2

The only know user of this would be wine (that I’m aware of) see https://github.com/wine-mirror/wine/commit/0185ee5d99e8dca2c69d61ba0c0e00256beaf1b5

Bypassing brew shims for CC & LD allowed me to compile wine-7.0 (wine64 only) see https://github.com/Gcenx/homebrew-wine/actions/runs/1853481245

As upstream is working on allowing 32Bit windows binaries to run without 32Bit host libraries that would be nice to eventually restore a wine formula in homebrew-core

**Comment by carlocab** [comment, marked as answer]
The difficulty is that setting `-mmacosx-version-min=10.8` or earlier causes errors when using `clang++`:
```
❯ echo 'int main() {}' | clang++ -xc++ - -mmacosx-version-min=10.8
clang: warning: include path for libstdc++ headers not found; pass '-stdlib=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
clang: warning: libstdc++ is deprecated; move to libc++ with a minimum deployment target of OS X 10.9 [-Wdeprecated]
ld: library not found for -lstdc++
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
How does Wine avoid this error? We could maybe pass `-stdlib=libc++` instead, but making this change could have lots of unintended consequences.

**Reply by Gcenx** [reply]
Only `wine-preloader` needs this to be set all other components (not compiled with mingw) will use what ever is set. Thought I don’t believe clang++ is even used for any actual wine components.

I did see some shim settings have overrides, I’d think as this is lightly uncommon maybe an override would be more viable then globally changing the min target. So if the override is set just skip injecting `-mmacosx-version-min=`

**Reply by carlocab** [reply]
There are overrides, but this isn't a good candidate for one, I'm afraid. Multiple formulae should require the override before we consider adding one.

**Reply by Gcenx** [reply]
I thought that might be the case so asking here instead of opening an issue directly, thanks for confirming :)

**Reply by carlocab** [reply]
Sure. It might be more viable for Wine to stop needing `-no_new_main` on macOS -- this flag isn't even documented in `man ld` anymore, and it requiring `-mmacosx-version-min=10.7` tells me that this is likely a deprecated flag that could go away in a future version of `ld`.

I won't be surprised if this flag isn't even recognised at all on ARM macOS (or doesn't do what you expect it to on the ARM slice of a Mach-O file).

**Reply by carlocab** [reply]
Though, looking at https://github.com/wine-mirror/wine/commit/0185ee5d99e8dca2c69d61ba0c0e00256beaf1b5:

It seems like `-no_new_main` is not required to build Wine? You won't be able to build the preloader, but it seems like something will still work without it.

**Reply by Gcenx** [reply]
While that's true the resulting wine will be mostly useless as preloader is used to workaround how macOS allocates memory.

**Reply by carlocab** [reply]
I see. Are there any barriers to the preloader not needing `-no_new_main`, beyond (potentially very painful) refactoring?

**Reply by Gcenx** [reply]
Here's a small section from `preloader_mac.c`

```
/* Rosetta on Apple Silicon allocates memory starting at 0x100000000 (the 4GB line)
 * before the preloader runs, which prevents any nonrelocatable EXEs with that
 * base address from running.
 *
 * This empty linker section forces Rosetta's allocations (currently ~132 MB)
 * to start at 0x114000000, and they should end below 0x120000000.
 */
#if defined(__x86_64__)
__asm__(".zerofill WINE_4GB_RESERVE,WINE_4GB_RESERVE,___wine_4gb_reserve,0x14000000");
#endif

#ifndef LC_MAIN
#define LC_MAIN 0x80000028
struct entry_point_command
{
    uint32_t cmd;
    uint32_t cmdsize;
    uint64_t entryoff;
    uint64_t stacksize;
};
#endif

static struct wine_preload_info preload_info[] =
{
    /* On macOS, we allocate the low 64k area in two steps because PAGEZERO
     * might not always be available. */
#ifdef __i386__
    { (void *)0x00000000, 0x00001000 },  /* first page */
    { (void *)0x00001000, 0x0000f000 },  /* low 64k */
    { (void *)0x00010000, 0x00100000 },  /* DOS area */
    { (void *)0x00110000, 0x67ef0000 },  /* low memory area */
    { (void *)0x7f000000, 0x03000000 },  /* top-down allocations + shared heap + virtual heap */
#else  /* __i386__ */
    { (void *)0x000000010000, 0x00100000 },  /* DOS area */
    { (void *)0x000000110000, 0x67ef0000 },  /* low memory area */
    { (void *)0x00007ff00000, 0x000f0000 },  /* shared user data */
    { (void *)0x000100000000, 0x14000000 },  /* WINE_4GB_RESERVE section */
    { (void *)0x7ffd00000000, 0x01ff0000 },  /* top-down allocations + virtual heap */
#endif /* __i386__ */
    { 0, 0 },                            /* PE exe range set with WINEPRELOADRESERVE */
    { 0, 0 }                             /* end of list */
};
```

**Reply by Gcenx** [reply]
> I see. Are there any barriers to the preloader not needing `-no_new_main`, beyond (potentially very painful) refactoring?

I don't believe it's possible to simply refactor it as all these functions were move outside of the standard `wine` binary in staging around wine-2.0 and into upstream wine-4.0, this fixed many applications and games. Even a simple old title (Total Annihilation for example) fails to launch without preloader.

**Reply by carlocab** [reply]
> Here's a small section from `preloader_mac.c`

That looks like a really terrible (but nonetheless impressive) hack 🙂
